dashboard/15-final/app/lib/actions.ts

```diff
@@ -1,11 +1,65 @@
 'use server';
 
-export async function deleteInvoice(id: number) {
-  // TO DO: Add delete invoice logic
-  console.log('Delete invoice', id);
+import { z } from 'zod';
+import { sql } from '@vercel/postgres';
+import { revalidatePath } from 'next/cache';
+import { redirect } from 'next/navigation';
+
+const InvoiceSchema = z.object({
+  id: z.string(),
+  customerId: z.string(),
+  amount: z.coerce.number(),
+  status: z.enum(['pending', 'paid']),
+  date: z.string(),
+});
+const CreateInvoice = InvoiceSchema.omit({ id: true, date: true });
+const UpdateInvoice = InvoiceSchema.omit({ date: true });
+const DeleteInvoice = InvoiceSchema.pick({ id: true });
+
+export async function createInvoice(formData: FormData) {
+  const { customerId, amount, status } = CreateInvoice.parse({
+    customerId: formData.get('customerId'),
+    amount: formData.get('amount'),
+    status: formData.get('status'),
+  });
+
+  const amountInCents = amount * 100;
+  const date = new Date().toISOString().split('T')[0];
+
+  await sql`
+    INSERT INTO invoices (customer_id, amount, status, date)
+    VALUES (${customerId}, ${amountInCents}, ${status}, ${date})
+  `;
+
+  revalidatePath('/dashboard/invoices');
+  redirect('/dashboard/invoices');
 }
 
-export async function addOrUpdateInvoice(formData: FormData) {
-  // TO DO: Add create/update invoice logic
-  console.log('Edit Invoice');
+export async function updateInvoice(formData: FormData) {
+  const { id, customerId, amount, status } = UpdateInvoice.parse({
+    id: formData.get('id'),
+    customerId: formData.get('customerId'),
+    amount: formData.get('amount'),
+    status: formData.get('status'),
+  });
+
+  const amountInCents = amount * 100;
+
+  await sql`
+    UPDATE invoices
+    SET customer_id = ${customerId}, amount = ${amountInCents}, status = ${status}
+    WHERE id = ${id}
+  `;
+
+  revalidatePath('/dashboard/invoices');
+  redirect('/dashboard/invoices');
+}
+
+export async function deleteInvoice(formData: FormData) {
+  const { id } = DeleteInvoice.parse({
+    id: formData.get('id'),
+  });
+
+  await sql`DELETE FROM invoices WHERE id = ${id}`;
+  revalidatePath('/dashboard/invoices');
```

Comment:
Why no `redirect` here vs the other mutations?

Reply:
Delete is on the same page as the invoices table, so we don't have to redirect.